### ISSUE TITLE
Complains about JAVA_HOME env variable not being set even though it is

### ISSUE THREAD

**Issue by PureTryOut** [fedora-java/javapackages-bootstrap#160]
I'm trying to get this running on a different distribution (Alpine Linux).
However using the `mvn_build.py` script from `javapackages-tools` to build it fails complaining that JAVA_HOME isn't set even though it is.

```
Executing: xmvn --batch-mode --offline -P\!quality verify org.fedoraproject.xmvn:xmvn-mojo:install org.fedoraproject.xmvn:xmvn-mojo:builddep
['xmvn', '--batch-mode', '--offline', '-P\\!quality', 'verify', 'org.fedoraproject.xmvn:xmvn-mojo:install', 'org.fedoraproject.xmvn:xmvn-mojo:builddep']
The JAVA_HOME environment variable is not defined correctly,
this environment variable is needed to run this program.
```

If I echo the variable just before it's set to `/usr/lib/jvm/java-21-openjdk` which is the right location. grep'ing through the sources I can not find where this error message is defined and thus what is complaining and why it can't find this variable that _does_ exist.

**Comment by PureTryOut** [comment]
Ah, for some reason that script has `export JAVA_HOME=/usr/lib/jvm/java-21` at the top, which is wrong. I don't see what part of the process adds that.
Interestingly enough the script at the version being used is a bit different, it has https://github.com/apache/maven/blob/8e8579a9e76f7d015ee5ec7bfcdc97d260186937/apache-maven/src/bin/mvn#L31-L45 (I'm linking to Maven 3.9.9 here) which the script installed locally doesn't have and it doesn't have that JAVA_HOME export.
Otherwise it seems to be the same.

**Comment by mizdebsk** [comment]
> **EDIT:** Ah, found it, it's https://github.com/fedora-java/javapackages-bootstrap/blob/master/patches/maven/0002-Bind-to-OpenJDK-21-for-runtime.patch. That's the Fedora specific part I suppose. I'll patch that.

Right, that explains it. That patch can be simply removed. After that don't forget to re-run  `./mbi.sh prep maven` to re-create `downstream/maven` based on `upstream/maven` and `patches/maven`.

**Comment by PureTryOut** [comment]
Thanks!

I have xmvn built now but now it's failing to install. Sorry to bother you with this all :see_no_evil: 
```
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-parent:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-api:jar:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-api:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-core:jar:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-core:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-connector:jar:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-connector:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-mojo:jar:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-mojo:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-tools:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-resolve:jar:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-resolve:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-subst:jar:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-subst:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-install:jar:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn-install:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn:pom:4.3.0
[INFO] Installing artifact org.fedoraproject.xmvn:xmvn:tar.gz:bin:4.3.0
[ERROR] Artifact installation failed
org.fedoraproject.xmvn.tools.install.ArtifactInstallationException: Installation repository is incapable of holding artifact org.fedoraproject.xmvn:xmvn:tar.gz:bin:SYSTEM
        at org.fedoraproject.xmvn.tools.install.impl.DefaultArtifactInstaller.install(DefaultArtifactInstaller.java:137)
        at org.fedoraproject.xmvn.tools.install.impl.DefaultInstaller.installArtifact(DefaultInstaller.java:227)
        at org.fedoraproject.xmvn.tools.install.impl.DefaultInstaller.install(DefaultInstaller.java:346)
        at org.fedoraproject.xmvn.tools.install.cli.InstallerCli.run(InstallerCli.java:58)
        at org.fedoraproject.xmvn.tools.install.cli.InstallerCli.doMain(InstallerCli.java:93)
        at org.fedoraproject.xmvn.tools.install.cli.InstallerCli.main(InstallerCli.java:105)
```

I'm calling `xmvn-install -R .xmvn-reactor -n xmvn -d /path/to/install/location` (removing `-d` and using the default doesn't help). Looking through the source a bit it goes wrong at https://github.com/fedora-java/xmvn/blob/master/xmvn-tools/xmvn-install/src/main/java/org/fedoraproject/xmvn/tools/install/impl/DefaultArtifactInstaller.java#L125, but looking at `getPrimaryArtifactPath()` I'm not sure why that would return `null` here, but then again I don't really know this codebase.

**Comment by mizdebsk** [comment]
> Sorry to bother you with this all 🙈

No problem, I'm happy to help.

> org.fedoraproject.xmvn.tools.install.ArtifactInstallationException: Installation repository is incapable of holding artifact org.fedoraproject.xmvn:xmvn:tar.gz:bin:SYSTEM

This indicates missing or incorrect configuration. XMvn uses configuration in XML format.
The format is documented at https://fedora-java.github.io/xmvn/config.html
XMvn loads it from various places: https://fedora-java.github.io/xmvn/configuration.html
Generally it tries to follow XDG specification.
Configuration used in Fedora and RHEL is available at https://github.com/fedora-java/javapackages/blob/master/configs/configuration.xml
The reason the configuration is part of javapackages project and not xmvn project is that it is distro-specific.

In this particular case, the build produced Maven artifact with groupId=org.fedoraproject.xmvn, artifactId=xmvn, extension=tar.gz, classifier=bin, version=SYSTEM. None of `<repositories>` described in the configuration is capable of holding this artifact - repository "base-raw-pom" holds artifacts with extension=pom, base-jar with extension=jar or extension=zip and so on, but none of them is expected to store tar.gz artifacts.

However there is a global packaging rule:
```
  <artifactManagement>
    <rule>
      <optional>true</optional>
      <artifactGlob>
	<classifier>*?</classifier>
      </artifactGlob>
      <targetPackage>__noinstall</targetPackage>
    </rule>
  </artifactManagement>
```
It basically says that any artifact with non-empty classifier shouldn't be installed by default.

In Fedora we are calling `mvn_package.py ::tar.gz: __noinstall` which configures XMvn not to install .tar.gz artifacts.
Example logs are at: https://kojipkgs.fedoraproject.org//packages/xmvn/4.3.0/2.fc42/data/logs/noarch/build.log
Related documentation is at https://docs.fedoraproject.org/en-US/java-packaging-howto/#_assignment_of_the_maven_artifacts_to_the_subpackages
The `%mvn_package` RPM macro corresponds to mvn_package.py Python script.

To summarize, the issue can be fixed by putting appropriate XMvn configuration in place.
I hope that helps.

**Comment by PureTryOut** [comment]
> In Fedora we are calling mvn_package.py ::tar.gz: __noinstall which configures XMvn not to install .tar.gz artifacts.
It's needed because we are also calling mvn_package.py ':{xmvn,xmvn-connector}' xmvn which says that any artifact with artifactId=xmvn should be installed into package "xmvn", overriding default rule mentioned above.

> The %mvn_package RPM macro corresponds to mvn_package.py Python script.

Yes, I figured that already so I am calling the same things (I'm taking inspiration from the Fedora packaging).

A snippet of the build script that has this error for me:

```sh
prepare() {
	default_prepare

	export JAVA_HOME=/usr/lib/jvm/java-21-openjdk

	mvn_package ::tar.gz __noinstall
	mvn_package ":{xmvn,xmvn-connector}" xmvn
	mvn_package ":xmvn-{api,core,parent}" core
	mvn_package ":xmvn-mojo" mojo
	mvn_package ":xmvn-{install,resolve,subst,tools}" tools

	# Don't put Class-Path attributes in manifests
	pom_remove_plugin :maven-jar-plugin xmvn-tools

	pom_remove_plugin :maven-dependency-plugin
}

build() {
	mvn_build -j -- -P\\!quality
}

package() {
	xmvn-install -R .xmvn-reactor -n $pkgname -d "$pkgdir"
}
```

`mvn_package`, `pom_remove_plugin` and `mvn_build` are basically the same as the Fedora macro's, they just call the Python scripts.

**Comment by mizdebsk** [comment]
Please attach somewhere xmvn-install logs with debugging enabled, `xmvn-install -X`.

> I do wonder, are `:{xmvn,xmvn-connector}` supposed to expand to 2 separate calls `:xmvn` and `:xmvn-connector` through Bash?
No, XMvn expands them by itself.

**Comment by PureTryOut** [comment]
Oh, ugh, I found the problem.

Rather than `mvn_package.py ::tar.gz: __noinstall` I called `mvn_package.py ::tar.gz __noinstall`. Can you spot the difference? :sweat_smile: 
I have a finished build now :tada:  Let's see if I can package some Java things fully offline without vendoring now.

**Comment by mizdebsk** [comment]
> Can you spot the difference? 😅

I didn't notice it either

> I have a finished build now 🎉 Let's see if I can package some Java things fully offline without vendoring now.

I'm glad it worked for you. Good luck with the packaging.

I'll happily accept patches that improve portability, remove bashism, make the scripts work with busybox etc.

**Comment by PureTryOut** [comment]
So I just had a fully bootstrapped path working and then I cleared out all packages to try the whole build from scratch again and now it fails building xmvn.

```
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] XMvn Parent                                                        [pom]
[INFO] XMvn API                                                           [jar]
[INFO] XMvn Core                                                          [jar]
[INFO] XMvn Connector for Maven Resolver                                  [jar]
[INFO] XMvn MOJO                                                 [maven-plugin]
[INFO] XMvn Tools                                                         [pom]
[INFO] XMvn Resolver                                                      [jar]
[INFO] XMvn Subst                                                         [jar]
[INFO] XMvn Install                                                       [jar]
[INFO] XMvn                                                               [pom]
[WARNING] Plugin not found in any plugin repository: org.fedoraproject.xmvn:xmvn-mojo
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for XMvn 4.3.0:
[INFO] 
[INFO] XMvn Parent ........................................ SKIPPED
[INFO] XMvn API ........................................... SKIPPED
[INFO] XMvn Core .......................................... SKIPPED
[INFO] XMvn Connector for Maven Resolver .................. SKIPPED
[INFO] XMvn MOJO .......................................... SKIPPED
[INFO] XMvn Tools ......................................... SKIPPED
[INFO] XMvn Resolver ...................................... SKIPPED
[INFO] XMvn Subst ......................................... SKIPPED
[INFO] XMvn Install ....................................... SKIPPED
[INFO] XMvn ............................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.183 s
[INFO] Finished at: 2024-11-13T20:55:40Z
[INFO] ------------------------------------------------------------------------
[ERROR] Error resolving version for plugin 'org.fedoraproject.xmvn:xmvn-mojo' from the repositories [local (/home/bart/.m2/repository), central (https://repo.maven.apache.org/maven2)]: Plugin not found in any plugin repository -> [Help 1]
```

Build order being xmvn -> maven -> xmvn-stage0 -> javapackages-tools-stage0 -> javapackages-bootstrap.
xmvn-stage0 does have `/usr/share/java/xmvn/xmvn-mojo.jar` but it seems the build system doesn't even look there. Not sure how I got this working earlier :thinking:

**Comment by mizdebsk** [comment]
First, try running `mvn_build` with option `-X`, which will enable some debugging info about artifact resolution.

**Comment by PureTryOut** [comment]
Good call, that's an useful option.

It reveals the following which seems relevant:

```
[DEBUG] Message scheme: plain
[DEBUG] Reading global settings from /usr/share/maven-3/conf/settings.xml
[DEBUG] Reading user settings from /home/bart/.m2/settings.xml
[DEBUG] Reading global toolchains from /usr/share/maven-3/conf/toolchains.xml
[DEBUG] Reading user toolchains from /home/bart/.m2/toolchains.xml
[DEBUG] Using local repository at /home/bart/.m2/repository
[DEBUG] Using manager EnhancedLocalRepositoryManager with priority 10.0 for /home/bart/.m2/repository
```

Except for "Using local repository at <wrong path>" that seems alright, but `/usr/share/maven-3/conf/toolchains.xml` doesn't exist.

Comparing this to building `xmvn-stage0` which uses `javapackages-bootstrap` it prints this instead:

```
[DEBUG] Reading global settings from /usr/share/xmvn/conf/settings.xml
[DEBUG] Reading user settings from /home/bart/.m2/settings.xml
[DEBUG] Reading global toolchains from /usr/share/xmvn/conf/toolchains.xml
[DEBUG] Reading user toolchains from /home/bart/.m2/toolchains.xml
[DEBUG] Using local repository at /home/bart/.m2/repository
[DEBUG] Using manager EnhancedLocalRepositoryManager with priority 10.0 for /home/bart/.m2/repository
```

Here `/usr/share/xmvn/conf/toolchains.xml` does exist. Could that be the critical difference? Interestingly enough it still says it's using the local repository from my home directory...

For reference, the complete build log of the failing xmvn:

```
Executing: xmvn --batch-mode --offline --debug -Dxmvn.debug -P\!quality verify org.fedoraproject.xmvn:xmvn-mojo:install org.fedoraproject.xmvn:xmvn-mojo:builddep
['xmvn', '--batch-mode', '--offline', '--debug', '-Dxmvn.debug', '-P\\!quality', 'verify', 'org.fedoraproject.xmvn:xmvn-mojo:install', 'org.fedoraproject.xmvn:xmvn-mojo:builddep']
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/share/maven-3/lib/maven-slf4j-provider-3.9.9.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/share/maven-3/lib/slf4j-api-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.MavenSimpleLoggerFactory]
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/share/maven-3/lib/maven-slf4j-provider-3.9.9.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/share/maven-3/lib/slf4j-api-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.MavenSimpleLoggerFactory]
Apache Maven 3.9.9
Maven home: /usr/share/maven-3
Java version: 21.0.5, vendor: Alpine, runtime: /usr/lib/jvm/java-21-openjdk
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.6.60-0-lts", arch: "amd64", family: "unix"
[DEBUG] Created new class realm maven.api
[DEBUG] Importing foreign packages into class realm maven.api
[DEBUG]   Imported: javax.annotation.* < plexus.core
[DEBUG]   Imported: javax.annotation.security.* < plexus.core
[DEBUG]   Imported: javax.inject.* < plexus.core
[DEBUG]   Imported: org.apache.maven.* < plexus.core
[DEBUG]   Imported: org.apache.maven.artifact < plexus.core
[DEBUG]   Imported: org.apache.maven.classrealm < plexus.core
[DEBUG]   Imported: org.apache.maven.cli < plexus.core
[DEBUG]   Imported: org.apache.maven.configuration < plexus.core
[DEBUG]   Imported: org.apache.maven.exception < plexus.core
[DEBUG]   Imported: org.apache.maven.execution < plexus.core
[DEBUG]   Imported: org.apache.maven.execution.scope < plexus.core
[DEBUG]   Imported: org.apache.maven.graph < plexus.core
[DEBUG]   Imported: org.apache.maven.lifecycle < plexus.core
[DEBUG]   Imported: org.apache.maven.model < plexus.core
[DEBUG]   Imported: org.apache.maven.monitor < plexus.core
[DEBUG]   Imported: org.apache.maven.plugin < plexus.core
[DEBUG]   Imported: org.apache.maven.profiles < plexus.core
[DEBUG]   Imported: org.apache.maven.project < plexus.core
[DEBUG]   Imported: org.apache.maven.reporting < plexus.core
[DEBUG]   Imported: org.apache.maven.repository < plexus.core
[DEBUG]   Imported: org.apache.maven.rtinfo < plexus.core
[DEBUG]   Imported: org.apache.maven.settings < plexus.core
[DEBUG]   Imported: org.apache.maven.toolchain < plexus.core
[DEBUG]   Imported: org.apache.maven.usability < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.* < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.authentication < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.authorization < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.events < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.observers < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.proxy < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.repository < plexus.core
[DEBUG]   Imported: org.apache.maven.wagon.resource < plexus.core
[DEBUG]   Imported: org.codehaus.classworlds < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.* < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.classworlds < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.component < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.configuration < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.container < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.context < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.lifecycle < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.logging < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.personality < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.util.xml.Xpp3Dom < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.util.xml.pull.XmlPullParser < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.util.xml.pull.XmlPullParserException < plexus.core
[DEBUG]   Imported: org.codehaus.plexus.util.xml.pull.XmlSerializer < plexus.core
[DEBUG]   Imported: org.eclipse.aether.* < plexus.core
[DEBUG]   Imported: org.eclipse.aether.artifact < plexus.core
[DEBUG]   Imported: org.eclipse.aether.collection < plexus.core
[DEBUG]   Imported: org.eclipse.aether.deployment < plexus.core
[DEBUG]   Imported: org.eclipse.aether.graph < plexus.core
[DEBUG]   Imported: org.eclipse.aether.impl < plexus.core
[DEBUG]   Imported: org.eclipse.aether.installation < plexus.core
[DEBUG]   Imported: org.eclipse.aether.internal.impl < plexus.core
[DEBUG]   Imported: org.eclipse.aether.metadata < plexus.core
[DEBUG]   Imported: org.eclipse.aether.repository < plexus.core
[DEBUG]   Imported: org.eclipse.aether.resolution < plexus.core
[DEBUG]   Imported: org.eclipse.aether.spi < plexus.core
[DEBUG]   Imported: org.eclipse.aether.transfer < plexus.core
[DEBUG]   Imported: org.eclipse.aether.util < plexus.core
[DEBUG]   Imported: org.eclipse.aether.version < plexus.core
[DEBUG]   Imported: org.fusesource.jansi.* < plexus.core
[DEBUG]   Imported: org.slf4j.* < plexus.core
[DEBUG]   Imported: org.slf4j.event.* < plexus.core
[DEBUG]   Imported: org.slf4j.helpers.* < plexus.core
[DEBUG]   Imported: org.slf4j.spi.* < plexus.core
[DEBUG] Populating class realm maven.api
[DEBUG] Created adapter factory; available factories [file-lock, rwlock-local, semaphore-local, noop]; available name mappers [discriminating, file-gav, file-hgav, file-static, gav, static]
[INFO] Error stacktraces are turned on.
[DEBUG] Message scheme: plain
[DEBUG] Reading global settings from /usr/share/maven-3/conf/settings.xml
[DEBUG] Reading user settings from /home/bart/.m2/settings.xml
[DEBUG] Reading global toolchains from /usr/share/maven-3/conf/toolchains.xml
[DEBUG] Reading user toolchains from /home/bart/.m2/toolchains.xml
[DEBUG] Using local repository at /home/bart/.m2/repository
[DEBUG] Using manager EnhancedLocalRepositoryManager with priority 10.0 for /home/bart/.m2/repository
[INFO] Scanning for projects...
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn:pom:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-parent:pom:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-api:jar:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-core:jar:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-connector:jar:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-mojo:maven-plugin:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging maven-plugin from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-tools:pom:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-resolve:jar:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-subst:jar:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.fedoraproject.xmvn:xmvn-install:jar:4.3.0: (none)
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[plexus.core, parent: null]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] XMvn Parent                                                        [pom]
[INFO] XMvn API                                                           [jar]
[INFO] XMvn Core                                                          [jar]
[INFO] XMvn Connector for Maven Resolver                                  [jar]
[INFO] XMvn MOJO                                                 [maven-plugin]
[INFO] XMvn Tools                                                         [pom]
[INFO] XMvn Resolver                                                      [jar]
[INFO] XMvn Subst                                                         [jar]
[INFO] XMvn Install                                                       [jar]
[INFO] XMvn                                                               [pom]
[DEBUG] Creating adapter using nameMapper 'gav' and factory 'rwlock-local'
[DEBUG] Could not find metadata org.fedoraproject.xmvn:xmvn-mojo/maven-metadata.xml in local (/home/bart/.m2/repository)
[DEBUG] Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the metadata org.fedoraproject.xmvn:xmvn-mojo/maven-metadata.xml has not been downloaded from it before
[WARNING] Plugin not found in any plugin repository: org.fedoraproject.xmvn:xmvn-mojo
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for XMvn 4.3.0:
[INFO] 
[INFO] XMvn Parent ........................................ SKIPPED
[INFO] XMvn API ........................................... SKIPPED
[INFO] XMvn Core .......................................... SKIPPED
[INFO] XMvn Connector for Maven Resolver .................. SKIPPED
[INFO] XMvn MOJO .......................................... SKIPPED
[INFO] XMvn Tools ......................................... SKIPPED
[INFO] XMvn Resolver ...................................... SKIPPED
[INFO] XMvn Subst ......................................... SKIPPED
[INFO] XMvn Install ....................................... SKIPPED
[INFO] XMvn ............................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.184 s
[INFO] Finished at: 2024-11-13T21:22:53Z
[INFO] ------------------------------------------------------------------------
[ERROR] Error resolving version for plugin 'org.fedoraproject.xmvn:xmvn-mojo' from the repositories [local (/home/bart/.m2/repository), central (https://repo.maven.apache.org/maven2)]: Plugin not found in any plugin repository -> [Help 1]
org.apache.maven.plugin.version.PluginVersionResolutionException: Error resolving version for plugin 'org.fedoraproject.xmvn:xmvn-mojo' from the repositories [local (/home/bart/.m2/repository), central (https://repo.maven.apache.org/maven2)]: Plugin not found in any plugin repository
    at org.apache.maven.plugin.version.internal.DefaultPluginVersionResolver.selectVersion (DefaultPluginVersionResolver.java:252)
    at org.apache.maven.plugin.version.internal.DefaultPluginVersionResolver.resolveFromRepository (DefaultPluginVersionResolver.java:153)
    at org.apache.maven.plugin.version.internal.DefaultPluginVersionResolver.resolve (DefaultPluginVersionResolver.java:100)
    at org.apache.maven.lifecycle.internal.MojoDescriptorCreator.resolvePluginVersion (MojoDescriptorCreator.java:260)
    at org.apache.maven.lifecycle.internal.MojoDescriptorCreator.getMojoDescriptor (MojoDescriptorCreator.java:223)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleTaskSegmentCalculator.calculateTaskSegments (DefaultLifecycleTaskSegmentCalculator.java:94)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleTaskSegmentCalculator.calculateTaskSegments (DefaultLifecycleTaskSegmentCalculator.java:76)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:84)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginVersionResolutionException
[DEBUG] Shutting down adapter factory; available factories [file-lock, rwlock-local, semaphore-local, noop]; available name mappers [discriminating, file-gav, file-hgav, file-static, gav, static]
[DEBUG] Shutting down 'file-lock' factory
[DEBUG] Shutting down 'rwlock-local' factory
[DEBUG] Shutting down 'semaphore-local' factory
[DEBUG] Shutting down 'noop' factory
```

**Comment by mizdebsk** [comment]
Looks like the difference is Maven vs XMvn?
Upstream Maven by itself is not capable of reading packages installed in the system.
For this you need extended Maven, XMvn. Only XMvn can resolve "local" artifacts from the system.

The right order in which packages are built is called "build plan": https://github.com/mizdebsk/mbici-config/blob/master/plan/bootstrap-all-rawhide.xml
The format of this file is documented at https://github.com/mizdebsk/mbici-workflow/?tab=readme-ov-file#build-plan
Basically:
- you first build javapackages-tools and javapackages-bootstrap, in parallel (phase b2)
- then all the other packages, ~100 of them (phase b3)
- at this point javapackages-bootstrap is no longer needed, all packages can be rebuild, except maven and xmvn (phase nb2)
- then you rebuild maven (phase nb3)
- and xmvn (phase nb4)

**Comment by PureTryOut** [comment]
> Looks like the difference is Maven vs XMvn?
> Upstream Maven by itself is not capable of reading packages installed in the system.
> For this you need extended Maven, XMvn. Only XMvn can resolve "local" artifacts from the system.

I got that yes, that's why the bootstrapped xmvn (xmvn-stage0) is present when building xmvn itself.

**Comment by PureTryOut** [comment]
Ah, found it, derp. I was not changing Maven home in my local adaption of https://src.fedoraproject.org/rpms/maven/blob/rawhide/f/0001-Adapt-mvn-script.patch#_55 with https://src.fedoraproject.org/rpms/xmvn/blob/rawhide/f/xmvn.spec#_181.
Fixed that, now it seems to look for the right Maven home.

However, now it fails because of a missing class, even though the jar that provides it is present.

```
Executing: xmvn --batch-mode --offline --debug -Dxmvn.debug -P\!quality verify org.fedoraproject.xmvn:xmvn-mojo:install org.fedoraproject.xmvn:xmvn-mojo:builddep
['xmvn', '--batch-mode', '--offline', '--debug', '-Dxmvn.debug', '-P\\!quality', 'verify', 'org.fedoraproject.xmvn:xmvn-mojo:install', 'org.fedoraproject.xmvn:xmvn-mojo:builddep']
Error: Could not find or load main class org.codehaus.plexus.classworlds.launcher.Launcher
Caused by: java.lang.ClassNotFoundException: org.codehaus.plexus.classworlds.launcher.Launcher
```

And indeed, just calling `xmvn --help` reveals the same error.
`xmvn-stage0` contains `/usr/share/xmvn/lib/plexus-classworlds.jar` just like `javapackages-bootstrap` does and I made sure that the Maven home is set to `/usr/share/xmvn` while calling through xmvn.
